File: modules/core/src/main/java/org/apache/ignite/binary/BinaryReader.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ignite.binary;

import java.math.BigDecimal;
import java.sql.Time;
import java.sql.Timestamp;
import java.util.Collection;
import java.util.Date;
import java.util.Map;
import java.util.UUID;

/**
 * Reader for binary objects used in {@link Binarylizable} implementations.
 * Useful for the cases when user wants a fine-grained control over serialization.
 * <p>
 * Note that Ignite never writes full strings for field or type names. Instead,
 * for performance reasons, Ignite writes integer hash codes for type and field names.
 * It has been tested that hash code conflicts for the type names or the field names
 * within the same type are virtually non-existent and, to gain performance, it is safe
 * to work with hash codes. For the cases when hash codes for different types or fields
 * actually do collide, Ignite provides {@link BinaryIdMapper} which
 * allows to override the automatically generated hash code IDs for the type and field names.
 */
public interface BinaryReader {
    /**
     * @param fieldName Field name.
     * @return Byte value.
     * @throws BinaryObjectException In case of error.
     */
    public byte readByte(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Short value.
     * @throws BinaryObjectException In case of error.
     */
    public short readShort(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Integer value.
     * @throws BinaryObjectException In case of error.
     */
    public int readInt(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Long value.
     * @throws BinaryObjectException In case of error.
     */
    public long readLong(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @throws BinaryObjectException In case of error.
     * @return Float value.
     */
    public float readFloat(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Double value.
     * @throws BinaryObjectException In case of error.
     */
    public double readDouble(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Char value.
     * @throws BinaryObjectException In case of error.
     */
    public char readChar(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Boolean value.
     * @throws BinaryObjectException In case of error.
     */
    public boolean readBoolean(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Decimal value.
     * @throws BinaryObjectException In case of error.
     */
    public BigDecimal readDecimal(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return String value.
     * @throws BinaryObjectException In case of error.
     */
    public String readString(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return UUID.
     * @throws BinaryObjectException In case of error.
     */
    public UUID readUuid(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Date.
     * @throws BinaryObjectException In case of error.
     */
    public Date readDate(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Timestamp.
     * @throws BinaryObjectException In case of error.
     */
    public Timestamp readTimestamp(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Time.
     * @throws BinaryObjectException In case of error.
     */
    public Time readTime(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Object.
     * @throws BinaryObjectException In case of error.
     */
    public <T> T readObject(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Byte array.
     * @throws BinaryObjectException In case of error.
     */
    public byte[] readByteArray(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Short array.
     * @throws BinaryObjectException In case of error.
     */
    public short[] readShortArray(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Integer array.
     * @throws BinaryObjectException In case of error.
     */
    public int[] readIntArray(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Long array.
     * @throws BinaryObjectException In case of error.
     */
    public long[] readLongArray(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Float array.
     * @throws BinaryObjectException In case of error.
     */
    public float[] readFloatArray(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Byte array.
     * @throws BinaryObjectException In case of error.
     */
    public double[] readDoubleArray(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Char array.
     * @throws BinaryObjectException In case of error.
     */
    public char[] readCharArray(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Boolean array.
     * @throws BinaryObjectException In case of error.
     */
    public boolean[] readBooleanArray(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Decimal array.
     * @throws BinaryObjectException In case of error.
     */
    public BigDecimal[] readDecimalArray(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return String array.
     * @throws BinaryObjectException In case of error.
     */
    public String[] readStringArray(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return UUID array.
     * @throws BinaryObjectException In case of error.
     */
    public UUID[] readUuidArray(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Date array.
     * @throws BinaryObjectException In case of error.
     */
    public Date[] readDateArray(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Timestamp array.
     * @throws BinaryObjectException In case of error.
     */
    public Timestamp[] readTimestampArray(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Time array.
     * @throws BinaryObjectException In case of error.
     */
    public Time[] readTimeArray(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Object array.
     * @throws BinaryObjectException In case of error.
     */
    public Object[] readObjectArray(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Collection.
     * @throws BinaryObjectException In case of error.
     */
    public <T> Collection<T> readCollection(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @param factory Collection factory.
     * @return Collection.
     * @throws BinaryObjectException In case of error.
     */
    public <T> Collection<T> readCollection(String fieldName, BinaryCollectionFactory<T> factory)
        throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Map.
     * @throws BinaryObjectException In case of error.
     */
    public <K, V> Map<K, V> readMap(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @param factory Map factory.
     * @return Map.
     * @throws BinaryObjectException In case of error.
     */
    public <K, V> Map<K, V> readMap(String fieldName, BinaryMapFactory<K, V> factory) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Value.
     * @throws BinaryObjectException In case of error.
     */
    public <T extends Enum<?>> T readEnum(String fieldName) throws BinaryObjectException;

    /**
     * @param fieldName Field name.
     * @return Value.
     * @throws BinaryObjectException In case of error.
     */
    public <T extends Enum<?>> T[] readEnumArray(String fieldName) throws BinaryObjectException;

    /**
     * Gets raw reader. Raw reader does not use field name hash codes, therefore,
     * making the format even more compact. However, if the raw reader is used,
     * dynamic structure changes to the binary objects are not supported.
     *
     * @return Raw reader.
     */
    public BinaryRawReader rawReader();
}

File: modules/platforms/cpp/binary/src/binary/binary_reader.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "ignite/binary/binary_reader.h"
#include "ignite/impl/binary/binary_reader_impl.h"

using namespace ignite::impl::binary;

namespace ignite
{
    namespace binary
    {
        BinaryReader::BinaryReader(BinaryReaderImpl* impl) : impl(impl)
        {
            // No-op.
        }
        
        int8_t BinaryReader::ReadInt8(const char* fieldName)
        {
            return impl->ReadInt8(fieldName);
        }

        int32_t BinaryReader::ReadInt8Array(const char* fieldName, int8_t* res, int32_t len)
        {
            return impl->ReadInt8Array(fieldName, res, len);
        }

        bool BinaryReader::ReadBool(const char* fieldName)
        {
            return impl->ReadBool(fieldName);
        }

        int32_t BinaryReader::ReadBoolArray(const char* fieldName, bool* res, int32_t len)
        {
            return impl->ReadBoolArray(fieldName, res, len);
        }

        int16_t BinaryReader::ReadInt16(const char* fieldName)
        {
            return impl->ReadInt16(fieldName);
        }

        int32_t BinaryReader::ReadInt16Array(const char* fieldName, int16_t* res, int32_t len)
        {
            return impl->ReadInt16Array(fieldName, res, len);
        }

        uint16_t BinaryReader::ReadUInt16(const char* fieldName)
        {
            return impl->ReadUInt16(fieldName);
        }

        int32_t BinaryReader::ReadUInt16Array(const char* fieldName, uint16_t* res, int32_t len)
        {
            return impl->ReadUInt16Array(fieldName, res, len);
        }

        int32_t BinaryReader::ReadInt32(const char* fieldName)
        {
            return impl->ReadInt32(fieldName);
        }

        int32_t BinaryReader::ReadInt32Array(const char* fieldName, int32_t* res, int32_t len)
        {
            return impl->ReadInt32Array(fieldName, res, len);
        }

        int64_t BinaryReader::ReadInt64(const char* fieldName)
        {
            return impl->ReadInt64(fieldName);
        }

        int32_t BinaryReader::ReadInt64Array(const char* fieldName, int64_t* res, int32_t len)
        {
            return impl->ReadInt64Array(fieldName, res, len);
        }

        float BinaryReader::ReadFloat(const char* fieldName)
        {
            return impl->ReadFloat(fieldName);
        }

        int32_t BinaryReader::ReadFloatArray(const char* fieldName, float* res, int32_t len)
        {
            return impl->ReadFloatArray(fieldName, res, len);
        }

        double BinaryReader::ReadDouble(const char* fieldName)
        {
            return impl->ReadDouble(fieldName);
        }

        int32_t BinaryReader::ReadDoubleArray(const char* fieldName, double* res, int32_t len)
        {
            return impl->ReadDoubleArray(fieldName, res, len);
        }

        Guid BinaryReader::ReadGuid(const char* fieldName)
        {
            return impl->ReadGuid(fieldName);
        }

        int32_t BinaryReader::ReadGuidArray(const char* fieldName, Guid* res, int32_t len)
        {
            return impl->ReadGuidArray(fieldName, res, len);
        }

        Date BinaryReader::ReadDate(const char * fieldName)
        {
            return impl->ReadDate(fieldName);
        }

        int32_t BinaryReader::ReadDateArray(const char * fieldName, Date * res, const int32_t len)
        {
            return impl->ReadDateArray(fieldName, res, len);
        }

        Timestamp BinaryReader::ReadTimestamp(const char * fieldName)
        {
            return impl->ReadTimestamp(fieldName);
        }

        int32_t BinaryReader::ReadTimestampArray(const char * fieldName, Timestamp * res, const int32_t len)
        {
            return impl->ReadTimestampArray(fieldName, res, len);
        }

        Time BinaryReader::ReadTime(const char* fieldName)
        {
            return impl->ReadTime(fieldName);
        }

        int32_t BinaryReader::ReadTimeArray(const char* fieldName, Time* res, const int32_t len)
        {
            return impl->ReadTimeArray(fieldName, res, len);
        }

        int32_t BinaryReader::ReadString(const char* fieldName, char* res, int32_t len)
        {
            return impl->ReadString(fieldName, res, len);
        }

        BinaryStringArrayReader BinaryReader::ReadStringArray(const char* fieldName)
        {
            int32_t size;

            int32_t id = impl->ReadStringArray(fieldName, &size);

            return BinaryStringArrayReader(impl, id, size);
        }

        BinaryEnumEntry BinaryReader::ReadBinaryEnum(const char* fieldName)
        {
            return impl->ReadBinaryEnum(fieldName);
        }

        CollectionType::Type BinaryReader::ReadCollectionType(const char* fieldName)
        {
            return impl->ReadCollectionType(fieldName);
        }

        int32_t BinaryReader::ReadCollectionSize(const char* fieldName)
        {
            return impl->ReadCollectionSize(fieldName);
        }

        BinaryRawReader BinaryReader::RawReader()
        {
            impl->SetRawMode();

            return BinaryRawReader(impl);
        }
    }
}
